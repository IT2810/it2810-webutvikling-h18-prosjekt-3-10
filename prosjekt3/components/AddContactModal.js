import React, { Component } from 'react';
import Colors from '../constants/Colors'
import {
    Modal,
    KeyboardAvoidingView,
    View,
    TextInput,
    StyleSheet,
    SafeAreaView,
    ScrollView,
} from 'react-native'
import {
    Avatar,
    Header,
    Button,
} from 'react-native-elements'

export default class AddContactModal extends Component {

    constructor(props) {
        super(props);
        this.state = {
            firstName: '',
            lastName: '',
            number: '',
            id: '',
        }
    }

    // Generates ID from first and last name combined with number
    // NOTE: This means it's possible (though unlikely) that two or more contacts can
    // share an ID.
    generateId = () => {
        const id = `${this.state.firstName}${this.state.lastName}${this.state.number}`
        return id
    }

    // Handles "Add contact"-button press
    handleAddContact = () => {
        const {
            firstName,
            lastName,
            number,
        } = this.state

        // Construct contact-object
        // Also trims spaces from both ends of names
        const contact = {
            firstName: firstName.trim(),
            lastName: lastName.trim(),
            id: this.generateId(),
            phoneNumbers: [{ number: number }]
        }

        // Reset modal state
        this.setState({
            firstName: '',
            lastName: '',
            number: '',
            id: '',
        })

        // Pass contact to save-method in parent screen/component
        this.props.onSave(contact)

        // Notify ready to close
        this.props.closeCallback()
    }

    render() {
        return (
            <Modal
                animationType="fade"
                transparent={false}
                visible={this.props.visible}
                onRequestClose={this.props.closeCallback}
            >
                <SafeAreaView style={styles.safeArea}>
                <KeyboardAvoidingView behavior="position">
                    <Header
                      outerContainerStyles={styles.headerOuterContainer}
                      rightComponent={{
                        size: 28,
                        icon: 'close',
                        color: 'black',
                        style: styles.headerBtn,
                        onPress: this.props.closeCallback
                      }}
                      centerComponent={{
                        text: 'Add contact',
                        style: styles.header,
                      }}
                      backgroundColor={Colors.headerBackground}
                    />
                    <View style={styles.modalContent}>
                        <Avatar // using react native elements - external libary
                            containerStyle={styles.avatar}
                            xlarge
                            rounded
                            source={require('../assets/images/profile.png')}
                            activeOpacity={0.7}
                        />

                        <View style={styles.nameInput}>
                            <TextInput style={styles.name} // input field for name
                                placeholder="First name"
                                value={this.state.firstName}
                                onChangeText={name => {
                                    this.setState({
                                        firstName: name,
                                    })
                                }}
                                underlineColorAndroid="transparent"
                            />
                        </View>

                        <View style={styles.nameInput}>
                            <TextInput style={styles.name} // input field for name
                                placeholder="Last name"
                                value={this.state.lastName}
                                onChangeText={name => {
                                    this.setState({
                                        lastName: name
                                    })
                                }}
                                underlineColorAndroid="transparent"
                            />
                        </View>

                        <View style={styles.numberInput}>
                            <TextInput
                                style={styles.number}
                                placeholder="Phone Number"
                                keyboardType='numeric'
                                onChangeText={(number) => {
                                    this.setState({
                                        number,
                                    })
                                }}
                                value={this.state.number}
                                maxLength={8}  // Limit for number of integers
                                underlineColorAndroid="transparent"
                            />
                        </View>
                        <View style={styles.buttonWrapper}>
                            <Button
                                onPress={this.handleAddContact}
                                buttonStyle={styles.addBtn}
                                title="Add"
                            />
                        </View>
                    </View>
                </KeyboardAvoidingView>
                </SafeAreaView>
            </Modal>
        )
    }
}

const styles = StyleSheet.create({
    safeArea: {
      flex: 1,
      backgroundColor: '#fff',
    },
    headerOuterContainer: {
      borderBottomWidth: 1,
      borderBottomColor: '#999',
    },
    header: {
      color: 'black',
      fontSize: 20,
      fontWeight: '600',
      fontVariant: ['small-caps'],
    },
    addBtn: {
      marginTop: 15,
      backgroundColor: Colors.btnBlue,
    },
    modalContent: {
        alignItems: "center",
    },
    avatar: {
        margin: 20,
    },

    name: {
        fontSize: 20,
        color: "#4d4d4d",
        fontWeight: '700',
        margin: 15,
    },
    number: {
        fontSize: 20,
        color: "#4d4d4d",
        fontWeight: '700',
        margin: 15,
    },
    nameInput: {
        width: 300,
    },
    numberInput: {
        width: 300,
    },
    buttonWrapper: {
        width: 200,
        marginBottom: 30,
    }
})
