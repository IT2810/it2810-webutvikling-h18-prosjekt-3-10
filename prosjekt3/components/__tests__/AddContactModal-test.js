import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import AddContactModal from '../AddContactModal';

describe('Snapshot testing', () => {
    jest.useFakeTimers();
    let tree = renderer.create(<AddContactModal />).toJSON();
    expect(tree).toMatchSnapshot();
})

describe('Unit testing', () => {
    let modal = renderer.create(<AddContactModal onSave={jest.fn()} closeCallback={jest.fn()}/>).getInstance();
    const exampleState = {
        firstName: "Hello",
        lastName: "World",
        number: "02020202",
        id: "1231231231",
    }
    const expectedInitialState = {
        firstName: '',
        lastName: '',
        number: '',
        id: '',
    }
    const { number } = exampleState
    const contact = {
        firstName: exampleState.firstName.trim(),
        lastName: exampleState.lastName.trim(),
        id: `${exampleState.firstName}${exampleState.lastName}${exampleState.number}`,
        phoneNumbers: [{number: number}]
    }

    it('should start with empty state', () => {
        expect(modal.state).toEqual(expectedInitialState);
    })

    it('should make contact, reset state and call onSave and close', () => {
        modal.setState(exampleState)
        expect(modal.state).toEqual(exampleState)
        modal.handleAddContact()
        expect(modal.state).toEqual(expectedInitialState)
        expect(modal.props.onSave).toBeCalledWith(contact)
        expect(modal.props.closeCallback).toBeCalled()
    })
});
