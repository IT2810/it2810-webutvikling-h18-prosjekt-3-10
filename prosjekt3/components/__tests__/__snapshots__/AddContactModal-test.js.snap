// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  transparent={false}
  visible={true}
>
  <View
    enabled={true}
    keyboardVerticalOffset={0}
    onLayout={[Function]}
  >
    <View
      style={
        Array [
          undefined,
          Object {
            "bottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#476DC5",
              "borderBottomColor": "#f2f2f2",
              "borderBottomWidth": 1,
              "height": 70,
              "padding": 15,
            },
            undefined,
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              },
              Object {
                "alignItems": "center",
              },
            ]
          }
        >
          <View
            style={
              Array [
                false,
                false,
                false,
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
            underlayColor="white"
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontSize": 32,
                  },
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              îŒ“
            </Text>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Object {
                  "color": "#fff",
                },
              ]
            }
          >
            Add contact
          </Text>
          <View
            style={
              Object {
                "height": 24,
                "width": 24,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <View
          activeOpacity={0.7}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "height": 150,
                "paddingBottom": 10,
                "paddingRight": 10,
                "paddingTop": 10,
                "width": 150,
              },
              Object {
                "borderRadius": 75,
              },
              Object {
                "margin": 20,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "stretch",
                  "backgroundColor": "rgba(0,0,0,0.2)",
                  "bottom": 0,
                  "flex": 1,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "borderRadius": 75,
                },
                undefined,
              ]
            }
          >
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": 150,
                    "width": 150,
                  },
                  Object {
                    "borderRadius": 75,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "width": 300,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="First name"
            style={
              Object {
                "color": "#4d4d4d",
                "fontSize": 20,
                "fontWeight": "700",
                "margin": 15,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "width": 300,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Last name"
            style={
              Object {
                "color": "#4d4d4d",
                "fontSize": 20,
                "fontWeight": "700",
                "margin": 15,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "width": 300,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            keyboardType="numeric"
            maxLength={8}
            onChangeText={[Function]}
            placeholder="Phone Number"
            style={
              Object {
                "color": "#4d4d4d",
                "fontSize": 20,
                "fontWeight": "700",
                "margin": 15,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginBottom": 30,
              "width": 200,
            }
          }
        >
          <View
            accessibilityComponentType="button"
            accessibilityTraits={
              Array [
                "button",
              ]
            }
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Add
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
